### PR TITLE
Add HoudiniSwap swap plugin

### PR DESCRIPTION
**Pull request by j0ntz**
### Technical Design Document

[stealth-send-swap.md](https://github.com/EdgeApp/edge-react-gui/blob/jon/stealth-send-swap/src/docs/stealth-send-swap.md)

### CHANGELOG

Does this branch warrant an entry to the CHANGELOG?

- [x] Yes
- [ ] No

### Dependencies

none (compiles against published edge-core-js; the synthetic-destination interaction is runtime-only and guarded)

### Description

[Asana task](https://app.asana.com/0/0/1216251688512498/f)

The HoudiniSwap swap plugin: privacy-routed CEX swaps through Houdini's v2 partner API, `initOptions: { apiKey, apiSecret }`. Two transport facts decide whether any call works at all, and neither is obvious. Auth is `Authorization: <key>:<secret>` with no `Bearer` prefix; every endpoint returns 402 without it. And the partner API is server-to-server and answers browser-origin requests with 403, so every call passes `corsBypass: 'always'` to route through the native fetch rather than the core WebView's.

**Route selection.** A request carrying `privacy: 'required'` takes `private` (multi-exchange) routes only, which is what makes a stealth flow private. Without it, `standard` routes are acceptable too, ranked below private. The distinction decides what a user can send: Houdini serves no private route under 25 USD but serves standard routes down to 10, so a plain swap-and-send between those figures is only possible on a standard route. A standard route still settles through Houdini, so the recipient never sees the sender's address, but it uses a single exchange leg that can relink the two sides, and the caller cannot inspect which route it got. That is why a privacy request declines rather than substituting. A `dex` route is the exception. A plain request that no private or standard route serves may take one, ranked below both, from a native coin on an EVM chain, priced on the send side, with no token approval to sign. Every Houdini route into Monad is a dex route, so this is what makes a Monad destination reachable at all. A privacy request never takes one: a dex route is an on-chain contract call the user's own wallet signs, which links the two sides in public.

**Dex orders.** A dex order names `addressFrom` and comes back with the call to sign rather than a deposit address, so the spend sends `metadata.value` to `metadata.to` with `metadata.data` as a hex memo, bounded by the requested amount exactly as a deposit is. A route Houdini would broadcast itself (`offChain`), or one missing any of those three fields, is declined and the next candidate gets its turn. After the broadcast `POST /dex/confirmTx` reports the txid, which is what starts the order on Houdini's side; a failure there is logged rather than reported as a failed send, since the funds have already left the wallet.

**Reverse quotes** (`quoteFor: 'to'`) map to `GET /quotes?amountType=receive&fixed=true`. The API prices exact-out on fixed-rate quotes alone, which its private routing does not serve, so they fall back to standard fixed-rate routes. Only that path sends `fixed=true`, so `isEstimate` is `!reverseQuote` rather than a constant, and a forward quote reports itself as an estimate whether its route is private or standard. Rejecting reverse quotes outright would break the flip input's guarantee semantics; emulating exact-out by inverting a forward quote loses the receive-side guarantee.

**Swap-to-address destinations.** A core-built synthetic destination wallet (id prefix `synthetic://`) skips the typed-address lookup, since it holds exactly one pasted, caller-validated address, and may expose destination memos through a `getMemos` method detected at runtime behind a local guarded type, so this package keeps compiling against published edge-core-js. The memo is forwarded as `destinationTag` on order creation. Memo chains XRP, XLM, ATOM, HBAR and RUNE are mapped; IBC-family chains (coreum, osmosis, axelar) stay unmapped because Houdini reports no `memoNeeded` flag and a permissive `^.*$` address validation for them.

**The chain table answers what Houdini calls a chain, not whether Houdini serves it.** Those are different questions with different lifetimes: the name is stable, while what is served changes whenever the provider adds or drops a native coin. Servedness is discovered at runtime by `resolveTokenId`, which declines with the same `SwapCurrencyError` the whitelist check raises, and which memoizes misses as well as hits so an unserved chain costs one call per ten-minute window rather than one per quote. The window cuts both ways: without it, a chain Houdini lists later in the session stays refused until the app restarts. A lookup the provider FAILED to answer is deliberately left uncached at all: a rate limit says nothing about whether a chain is served, and caching it would turn one bad minute into a chain that stays dead for the whole window. A non-OK `GET /tokens` therefore throws with its status rather than returning a miss, which the quote path would otherwise surface as a pair Houdini cannot route. The lookup itself trusts its query to scope the answer: the native query asks for `mainnet=true`, the catalogue's own definition of a chain's coin, and takes the row it gets back, while a contract token matches on its address alone. Re-testing rows only lost chains: TON's coin carries a contract-style address, and a `bitcoincash` query answers with rows on `bch`. Replaying that predicate against the live catalogue for every mapped chain (2026-09-10) resolves 35 of 38, the three misses being `celo`, `fantom` and `polkadot`, which have no mainnet native at all. `monad` (`MON`) and `robinhood` (`Robinhood`) are mapped; both resolve.

**Rate limits.** Houdini is an aggregator behind Cloudflare that allows one exchange per minute, and a 429 arriving where a quote was expected reads exactly like an unavailable pair. Every call goes through one wrapper that retries behind the `retryAfter` the API reports, and four things keep that budget honest:

- The local exponential cap bounds our own doubling only, never the provider's window. A 30s ceiling on Houdini's roughly 60s window meant the retry fired while still inside it, drew another 429, and spent the retries for nothing.
- A wait that would outlive what it is resending fails immediately as a rate limit. A quote id lives about a minute, so honoring a 60s window and re-POSTing the same id hangs the user for that minute and then reports an expired quote, blaming the wrong thing. The create call passes the candidate quote's own expiry into the wrapper.
- `validUntil` arrives as Unix seconds inside a string, which `new Date` reads as an invalid date, so the parse reads the number first or that guard can never fire.
- `getMaxSwappable` runs the quote function once to size the spend and the real quote runs it again, so the sizing pass builds its spend shape from the quote alone rather than creating a throwaway exchange, standing in the user's own refund address for the deposit address it does not have. Standing in the user's own address forces two more properties on that probe: its `spendInfo` sets `skipChecks: true`, without which an EVM engine rejects the spend-to-self match with `SpendToSelfError` and fails every max swap from an EVM wallet, and it clamps rather than throwing `SwapAboveLimitError`, since the probe deliberately quotes the full pre-fee balance and an above-limit balance still makes a max swap once the fee comes off. The route maximum is enforced on the real quote only.

**Amount safety.** Provider amounts arrive as JSON floats, so they reach `biggystring` through a decimal-string expansion that covers scientific notation at both ends; comparison and sorting go through `biggystring` too, since `String(smallFloat)` can produce notation a string comparison misreads. Rounding to whole atomic units carries a direction: a minimum rounds UP so the floor Edge enforces never lands below the provider's own, and a maximum, the receive amount and the deposit amount round DOWN so none is ever larger than what the provider honors. The deposit amount is a trust boundary as well: a `from` quote refuses an order whose `inAmount` exceeds what the user requested, before it becomes a signed spend. A reverse quote pins the receive side, so its order is bounded by the route it was created on: the smaller of that route's quoted send amount plus 1% for provider rounding, and its from-side ceiling. The quoted amount does the real work, since routes commonly publish `max: 9007199254740991`; a live fixed-rate order restated its quote at +0.012%.

**Cleaners.** The deposit tag uses `asOptionalBlank(asNumberString)` rather than `asOptional(asString)`: a numeric tag is the common shape on memo chains (the valid tag `0` included) and would take the whole order down against a string-only cleaner, while an empty string would become an empty `EdgeMemo` on the deposit. Every response is cleaned, and a cleaner failure logs the payload, which is the one error whose message otherwise says nothing about what arrived.

**Other behavior.** A fixed-rate route's static deposit address can be held by another live order (HTTP 409 `STATIC_DEPOSIT_IN_USE`, hit live during testing), so order creation falls through to the next-best in-range route; that case is recognized by the envelope's machine-readable `code`, read from the same single parse of the body that produces the error message, not by searching the response text, so a code appearing inside a human-readable message cannot trigger it. Forward limits (`min`/`max`) are from-side and reverse limits (`minOut`/`maxOut`) receive-side, and a reverse quote must also clear the route's from-side bounds with its own priced `amountIn`. `VALIDATION_ERROR` sets a generic top-level "Validation Failed" and puts the actionable text under `fields.<name>.message`, so field messages win over the top-level one; a specific top-level message with no fields surfaces unchanged. Zcash destinations use transparent addresses.

**Release ordering.** `EdgeSwapRequest.privacy` is added by [EdgeApp/edge-core-js#730](https://github.com/EdgeApp/edge-core-js/pull/730) and does not exist in the published `edge-core-js`, which is why `EdgeSwapRequestPlugin` declares it locally. That makes the ordering a release requirement rather than a preference: publish this package against a core that lacks the field and `privateOnly` reads `false` on every request, so a stealth send takes a standard route and reports no error. This package must not publish ahead of the core release that carries the field.

**Same-asset is allowed here, for private requests only.** Every other central plugin rejects a swap from an asset to itself through the shared `checkInvalidTokenIds`, which is right for a provider where it would be a no-op the user cannot have meant. Routing an asset to itself through a mixer is this provider's dominant flow, so the shared helper gained an `allowSameAsset` option that only Houdini passes, and only for a request carrying `privacy: 'required'`. A same-asset request without it is the same no-op and is declined. The blocked-token half of the helper applies either way. That is the first of the two commits here, so the shared change reviews on its own.

**Testing.** 130 mocha tests pass, `tsc` and eslint clean. 40 of those are this plugin's: 4 acceptance tests replaying fixtures recorded against the live API (forward BTC to ETH and ETH to USDC private swaps, a reverse BTC to ETH swap priced by the receive amount, and a synthetic memo-chain destination asserting the entered tag reaches the create-exchange body), and 36 offline behaviors driven from scripted local responses, including a coin whose row carries a contract-style address, a row naming its chain differently from the query (both fail on the previous predicate and pass on this one), the probe's `skipChecks` flag, the probe clamping an above-limit balance, the trust boundary refusing an inflated deposit amount, the rounding direction on both limits, a numeric and a blank deposit tag, the 409 fallthrough to the next route, a reverse quote's deposit bounded by the route's from-side ceiling and by its own quoted send amount (with an order that restates its quote within rounding accepted), and a same-asset request declined unless it asks for privacy. The offline half exists because a recorded fixture replays one canned answer per URL, which cannot express a SEQUENCE of statuses (the backoff needs 429 then 200) or a route mix the live API will not produce on demand (a pair offering transparent routes and no private one). The dex path has five offline cases: the route taken and `dex/confirmTx` posting the txid after broadcast, a quote needing a token approval declined, a non-EVM source declined, an off-chain route declined, and a transaction value above the request refused. In-app, live quotes through this plugin were exercised on the iOS simulator via the Stealth Send UI, through to executed private orders. A pasted Monad address from a Polygon wallet quoted 521.0386 MON for 124.88 POL and slid to the success scene; Houdini's order `2JrCoeLG36zo7pdRRWcwYn` records `isDex: true` and the source transaction hash this plugin confirmed. On the current head, a 12 USD Dash to Ethereum send through a standard route (1 DASH = 0.02090847 ETH) reached the success scene; the same pair priced by the receive side sent 0.22986803 DASH for a guaranteed 0.0047724 ETH on a fixed quote whose route published `max: 9007199254740991`, so the quoted-amount bound was the one in force; and a same-asset private TRX send (77.054 TRX, `privacy: 'required'` through core) reached the success scene.

In the app on the iOS simulator, with this branch and edge-core-js#730 linked into edge-react-gui#6066: a Stealth Swap of 1,043.63 S (27.99 USD) from Sonic to 20.2316 TON through the private route, driven to the success scene. TON could not resolve a token id on the previous lookup, so no quote into it was possible before.

---

> [!NOTE]
> **Medium Risk**
> New third-party swap path handles real deposits, privacy routing, and rate-limit/order semantics; mistakes could mis-quote, leak route type, or stall swaps, but changes are isolated to a new plugin with broad test coverage.
> 
> **Overview**
> Adds a **HoudiniSwap** central swap plugin wired into the plugin registry, with an Edge→Houdini chain mapping table and partner API integration (`apiKey`/`apiSecret`, CORS-bypass fetch).
> 
> Quotes resolve Houdini token IDs with session memoization (including misses), rank **private** over **standard** routes, honor **`privacy: 'required`** by declining when only transparent routes exist, and never use dex routes. Forward quotes are estimates; reverse **`quoteFor: 'to`** quotes use fixed receive pricing. Swap-to-address synthetic destinations forward **destination tags**; max-quote sizing skips creating an exchange to avoid the 1/min exchange budget. Rate limits retry per **`retryAfter`** without waiting past quote expiry.
> 
> Shared helpers gain **`checkInvalidTokenIds`** **`allowSameAsset`** (used only here for same-asset mixer flows) and **`EdgeSwapRequestPlugin.privacy`**. Acceptance and offline scripted tests plus disk fixtures cover quoting, orders, and edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 2dd7a791f0270e7f3b8204eab3ecc09d2aa7d2e9. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### Test evidence

<table>
<tr>
<td rowspan="2" width="170" valign="top"><a href="https://github.com/EdgeApp/edge-exchange-plugins/pull/469/commits/384ebb906c191ebe70d0d6d4eea60abb30e30060"><code>384ebb9</code></a><br><sub>Add the HoudiniSwap swap plugin</sub></td>
<td width="210" valign="top" align="center"><a href="https://raw.githubusercontent.com/EdgeApp/edge-dev-agents/agent-pr-assets/assets/edge-exchange-plugins/pr-469/20260702-200825-agent-proof-1216251688512498-01-stealth-send-toggle.png"></a><br><sub>agent proof 1216251688512498 01 stealth send toggle</sub></td>
<td width="210" valign="top" align="center"><a href="https://raw.githubusercontent.com/EdgeApp/edge-dev-agents/agent-pr-assets/assets/edge-exchange-plugins/pr-469/20260702-200825-agent-proof-1216251688512498-06-stealth-swap-houdini-only.png"></a><br><sub>agent proof 1216251688512498 06 stealth swap houdini only</sub></td>
<td width="210" valign="top" align="center"><a href="https://raw.githubusercontent.com/EdgeApp/edge-dev-agents/agent-pr-assets/assets/edge-exchange-plugins/pr-469/20260911-151311-agent-proof-1216251688512498-29-monad-dex-quote-armed.png"></a><br><sub>monad dex quote armed</sub></td>
</tr>
<tr>
<td width="210" valign="top" align="center"><a href="https://raw.githubusercontent.com/EdgeApp/edge-dev-agents/agent-pr-assets/assets/edge-exchange-plugins/pr-469/20260911-151311-agent-proof-1216251688512498-30-monad-dex-success.png"></a><br><sub>monad dex success</sub></td>
<td width="210" valign="top" align="center"><a href="https://raw.githubusercontent.com/EdgeApp/edge-dev-agents/agent-pr-assets/assets/edge-exchange-plugins/pr-469/20260911-151311-agent-proof-1216251688512498-34-monad-dex-txdetails.png"></a><br><sub>monad dex txdetails</sub></td>
<td width="210"></td>
</tr>
</table>

### CODE REVIEW

**Comment by j0ntz** [comment]
## 📸 Test evidence (live quotes through the plugin on the iOS sim)

**agent proof 1216251688512498 01 stealth send toggle**

agent proof 1216251688512498 01 stealth send toggle

**agent proof 1216251688512498 06 stealth swap houdini only**

agent proof 1216251688512498 06 stealth swap houdini only

_Captured by the agent's in-app test run (build-and-test)._

File: src/swap/central/houdini.ts [hunk 1]
```diff
@@ -0,0 +1,667 @@
...
+      networkFeeOption: 'high',
+      assetAction: {
+        assetActionType: 'swap'
+      },
+      savedAction: {
+        actionType: 'swap',
+        swapInfo,
+        orderId: order.houdiniId,
+        orderUri: orderUri + order.houdiniId,
+        isEstimate: false,
```
[the remark this answers is not in the document]
Comment (j0ntz) [reply src/swap/central/houdini.ts]: Correct. Fixed.

Only the exact-out path sends `fixed=true`, and that is the only path the provider serves fixed rates on, so `isEstimate` is now `!reverseQuote` rather than a hardcoded `false`. A forward quote reports itself as an estimate, private or standard, which is what it is. `makeSwapPluginQuote` reads `isEstimate` off the saved action, so the quote and the transaction details agree without a second change.

File: src/swap/central/houdini.ts [hunk 2]
```diff
@@ -0,0 +1,667 @@
...
+      checkWhitelistedMainnetCodes(
+        MAINNET_CODE_TRANSCRIPTION,
+        request,
+        swapInfo
+      )
+
+      const newRequest = await getMaxSwappable(fetchSwapQuoteInner, request)
+      const swapOrder = await fetchSwapQuoteInner(newRequest)
+      return await makeSwapPluginQuote(swapOrder)
+    }
```
[the remark this answers is not in the document]
Comment (j0ntz) [reply src/swap/central/houdini.ts:824]: Half right, and fixed on that half.

The blocked-token checks were genuinely missing: `checkInvalidTokenIds` is now called, so the shared `defaultInvalidCodes` list applies here like everywhere else.

The `isSameAsset` rejection is the half that cannot apply. Routing an asset to itself through a mixer is this provider's main flow, not a user mistake: it is roughly 60% of their traffic, they publish a per-asset `hasSelfPrivate` capability for it, and it is what the Stealth Send toggle does on a same-asset send. Adopting the guard unchanged would reject that outright.

Rather than inline a partial copy of the helper, the helper took an `allowSameAsset` option, which this plugin passes and no other caller sets. One implementation, and the exception is stated where it applies.

Verified: tsc clean, 37 tests passing, verify-repo PASSED, and the rebuilt bundle re-driven on the sim (a 32 TRX cross-asset send-to-address still arms at $10.31).

File: src/swap/central/houdini.ts [hunk 3]
```diff
@@ -0,0 +1,696 @@
...
+    if (!quoteResponse.ok) {
+      const text = await quoteResponse.text()
+      // Surface the API's own human-readable message when it carries one
+      // (e.g. "Amount is too low, minimum is 25 USD") instead of raw JSON:
+      const apiError = asHoudiniApiError(text)
+      throw new Error(
+        apiError != null
+          ? `HoudiniSwap: ${apiError.message}`
+          : `Houdini quotes returned ${quoteResponse.status}: ${text}`
+      )
```
[the remark this answers is not in the document]
Comment (j0ntz) [reply src/swap/central/houdini.ts:509]: Not taking this one, and the reason is worth recording.

The typed errors are already thrown wherever a native limit is knowable: when the API returns quotes carrying `min`/`max` (or `minOut`/`maxOut`), the in-range check below raises `SwapBelowLimitError` / `SwapAboveLimitError` with a real native amount and direction. That is the path that can be classified, and it is.

The 422 path this comment points at is different. The API answers `{"code":"AMOUNT_TOO_LOW","message":"Amount is too low, minimum is 5 USD"}` and nothing else. Those errors take a NATIVE amount, and the plugin has no fiat rate to turn 5 USD into one. Constructing one with a zero minimum would classify the error at the cost of the message: the GUI renders a zero-minimum below-limit error as a generic 'amount is below the min limit' string, which would delete the provider's own figure. Token-specific minimums above the shared floor arrive only through this path, and surfacing that exact number is behavior an operator round explicitly asked for.

So the choice is classification versus the specific number, and the number is worth more here. Recorded as a follow-up for whenever the API exposes the minimum in native units, at which point both are possible.

**Comment by j0ntz** [comment]
bugbot run

File: src/swap/central/houdini.ts [hunk 4]
```diff
@@ -0,0 +1,696 @@
...
+        ? `tokens?chain=${chain}&address=${contractAddress}&pageSize=100`
+        : `tokens?chain=${chain}&mainnet=true&pageSize=100`
+    const response = await fetchHoudini(query)
+    if (!response.ok) {
+      // Deliberately NOT cached. A rate limit or a server error says nothing
+      // about whether the chain is served, and caching it would turn one bad
+      // minute into a chain that stays dead for the rest of the session.
+      const text = await response.text()
+      log.warn('Houdini tokens lookup error:', text)
+      return undefined
```
[the remark this answers is not in the document]
Comment (j0ntz) [reply src/swap/central/houdini.ts]: Fixed in b3e1381. The lookup no longer returns a miss on a non-OK response: it throws `Houdini tokens returned <status>: <body>`, so the quote path cannot convert a provider outage into `SwapCurrencyError`. The failure is still not cached, so a recovered API resolves normally on the next quote. A new test asserts the thrown error names the status and is not a currency error.

File: src/swap/central/houdini.ts [hunk 5]
```diff
@@ -0,0 +1,698 @@
...
+      fromAddress,
+      toAddress,
+      toMemos
+    ] = await Promise.all([
+      resolveTokenId(fromMainnet, fromContractAddress),
+      resolveTokenId(toMainnet, toContractAddress),
+      getAddress(fromWallet, addressTypeMap[fromWallet.currencyInfo.pluginId]),
+      getAddress(toWallet, toAddressType),
+      getDestinationMemos(toWallet)
+    ])
```
[the remark this answers is not in the document]
Comment (j0ntz) [reply src/swap/central/houdini.ts:481]: Fixed in 6970169. The cache now holds the in-flight promise rather than the settled result, so concurrent askers share one `GET /tokens` call. A rejected lookup is deleted from the map, so a failed call still does not stick. New test asserts a same-asset quote produces exactly one lookup for that chain.

File: src/swap/central/houdini.ts [hunk 6]
```diff
@@ -0,0 +1,717 @@
...
+      // `retryAfter` is the floor the API asked for; the doubling on top of it
+      // keeps a burst from re-colliding at the moment the window reopens.
+      const baseMs =
+        rateLimit?.retryAfter == null
+          ? RATE_LIMIT_MIN_DELAY_MS
+          : rateLimit.retryAfter * 1000
+      const delayMs = Math.min(
+        Math.max(baseMs, RATE_LIMIT_MIN_DELAY_MS) * 2 ** attempt,
+        RATE_LIMIT_MAX_DELAY_MS
+      )
```
[the remark this answers is not in the document]
Comment (j0ntz) [reply src/swap/central/houdini.ts]: Fixed in 8467a7e. The cap now bounds only our own doubling; the API's `retryAfter` floor always survives it, so a 60s window waits 60s. Extracted the math into `rateLimitDelayMs` so it is directly testable, with a test pinning that a reported 60s window is never undercut and that the cap still bounds growth when no window is reported.

File: src/swap/central/houdini.ts [hunk 7]
```diff
@@ -0,0 +1,729 @@
...
+        allowSameAsset: true
+      })
+      checkWhitelistedMainnetCodes(
+        MAINNET_CODE_TRANSCRIPTION,
+        request,
+        swapInfo
+      )
+
+      const newRequest = await getMaxSwappable(fetchSwapQuoteInner, request)
+      const swapOrder = await fetchSwapQuoteInner(newRequest)
```
[the remark this answers is not in the document]
Comment (j0ntz) [reply src/swap/central/houdini.ts:822]: Fixed in 332fbbe. `fetchSwapQuoteInner` now takes a probe flag, forwarded by `getMaxSwappable` as its extra argument, and the sizing pass builds its spend shape from the quote alone instead of creating an exchange. It stands in the user's own refund address for the deposit address it does not have, which is on the from chain exactly as the real one would be, so fee estimation sees the same shape. New test asserts a max quote posts exactly one exchange.

File: src/swap/central/houdini.ts [hunk 8]
```diff
@@ -0,0 +1,729 @@
...
+type HoudiniQuote = ReturnType<typeof asHoudiniQuote>
+
+const asHoudiniQuotesResponse = asObject({
+  quotes: asArray(asMaybe(asHoudiniQuote))
+})
+
+/** The API's error envelope; `message` is human-readable. */
+const asHoudiniApiError = asMaybe(
+  asJSON(asObject({ message: asString }).withRest)
+)
```
[the remark this answers is not in the document]
Comment (j0ntz) [reply src/swap/central/houdini.ts:166]: Fixed in 332fbbe (same commit). The error cleaner now also reads `fields.<name>.message`, and a new `houdiniErrorMessage` helper prefers those over the top-level string, which is generic exactly when they are present. Both the quote and the create-exchange paths go through it. A specific top-level message with no fields ("Amount is too low, minimum is 25 USD") still surfaces unchanged. New test asserts a `VALIDATION_ERROR` reaches the user as "Quote has expired" rather than "Validation Failed".

File: src/swap/central/houdini.ts [hunk 9]
```diff
@@ -0,0 +1,791 @@
...
+
+      const delayMs = rateLimitDelayMs(attempt, rateLimit?.retryAfter)
+      log.warn(
+        `Houdini rate limited (${
+          rateLimit?.limit ?? '?'
+        }/window), retrying in ${delayMs}ms`
+      )
+      await snooze(delayMs)
+    }
+  }
```
[the remark this answers is not in the document]
Comment (j0ntz) [reply src/swap/central/houdini.ts:361]: Fixed in 234da7f. `fetchHoudini` now takes the expiry of whatever it is sending, and the create-exchange call passes the candidate quote's own `validUntil` (falling back to the documented 60s lifetime when the API omits it). A backoff that would land past that expiry fails as a rate limit immediately instead of sleeping through the window and reporting an expired quote. New test asserts a 60s `retryAfter` against a quote with 20s left returns a rate-limit error in under 5s and posts one exchange, not three.

File: src/swap/central/houdini.ts [hunk 10]
```diff
@@ -0,0 +1,820 @@
...
+ * when present; the constant is the documented default for when it is not.
+ */
+const QUOTE_LIFETIME_MS = 60000
+
+function quoteValidUntilMs(quote: HoudiniQuote): number {
+  const parsed =
+    quote.validUntil == null ? NaN : new Date(quote.validUntil).valueOf()
+  if (!isNaN(parsed)) return parsed
+  return Date.now() + QUOTE_LIFETIME_MS
+}
```
[the remark this answers is not in the document]
Comment (j0ntz) [reply src/swap/central/houdini.ts:247]: Confirmed against the recorded fixtures, which hold `"1783037880"`. Fixed in 491daab: the helper parses the number first (Unix seconds) and keeps the date parse as a fallback for an ISO form. New test pins that a seconds string round-trips to the right millisecond value, so the guard can actually fire.

File: src/swap/central/houdini.ts [hunk 11]
```diff
@@ -0,0 +1,844 @@
...
+  async function fetchTokenId(
+    chain: string,
+    contractAddress: string | undefined,
+    addressKey: string
+  ): Promise<string | undefined> {
+    const query =
+      contractAddress != null
+        ? `tokens?chain=${chain}&address=${contractAddress}&pageSize=100`
+        : `tokens?chain=${chain}&mainnet=true&pageSize=100`
+    const response = await fetchHoudini(query)
```
[the remark this answers is not in the document]
Comment (j0ntz) [reply src/swap/central/houdini.ts:504]: Fixed. The query now sends `addressKey` (the lowercased contract) instead of Edge's own spelling, so the wire form and the local match agree. Checksummed EVM metadata previously went out mixed-case, and a case-sensitive filter on the provider's side answers that with no tokens at all, which this plugin would report as a pair Houdini cannot route.

**Comment by j0ntz** [comment]
## 📸 Monad reached over a dex route, on device

**monad dex quote armed**

monad dex quote armed

**monad dex success**

monad dex success

**monad dex txdetails**

monad dex txdetails

_Captured by the agent's in-app test run (build-and-test)._
